### PR TITLE
feat(circleci): Add support for CircleCi

### LINKED ISSUE #1717: (circleci): add CircleCi support
I would like add support for generating CircleCi pipelines. We use CircleCi in our company quite heavily for CDK projects and it would help us to keep pipelines aligned and reduce the complexity of pipelines by generating those with Projen.

### PR DESCRIPTION
**Pull request by karlderkaefer**
Adding support to generate pipelines for CircleCi

fixes #1717 
---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by Chriscbr** [comment]
Hey @karlderkaefer - thanks for taking initiative on this! This is definitely a feature we want in projen -- but it might be a while before we're able to get to review this. I appreciate your patience 🙏

File: src/circleci/circleci.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+ * Circleci Class to manage `.circleci/config.yml`.
+ * Check test `circleci.test.ts` for more usage examples
+ *
+ * @see https://circleci.com/docs/2.0/configuration-reference/
+ */
+export class Circleci extends Component {
+  /**
+   * The yaml file for the Circleci pipeline
+   * */
+  public readonly file: YamlFile | undefined;
```
Comment (Chriscbr) [inline src/circleci/circleci.ts:84]: See other comment
```suggestion
  public readonly file: YamlFile;
```

File: src/circleci/circleci.ts [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+   * internal list of jobs
+   * @private
+   */
+  private readonly jobs: Job[];
+
+  /**
+   * create a Circleci Component within project
+   * @param project
+   * @param options
+   */
```
Comment (Chriscbr) [inline src/circleci/circleci.ts:110]: No need to have a docstring for the constructor
```suggestion
```

File: test/cirlceci/circleci.test.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    "Circleci Config already contains an orb named hello."
+  );
+
+  const snapshot = synthSnapshot(p);
+  const circleci = snapshot[".circleci/config.yml"];
+  expect(circleci).toMatchSnapshot();
+  expect(circleci).toContain("workflow1");
+  expect(circleci).toContain("world:2.0");
+  expect(circleci).not.toContain("world:3.0");
+  console.log(circleci);
```
Comment (Chriscbr) [inline test/cirlceci/circleci.test.ts:214]:
```suggestion
```

File: test/cirlceci/circleci.test.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  circle.addWorkflow({
+    identifier: "workflow3",
+    jobs: [{ identifier: "job3", context: ["context3"] }],
+  });
+  const snapshot = synthSnapshot(p);
+  const circleci = snapshot[".circleci/config.yml"];
+  expect(circleci).toMatchSnapshot();
+  expect(circleci).toContain("customName2");
+  expect(circleci).toContain("context3");
+  console.log(circleci);
```
Comment (Chriscbr) [inline test/cirlceci/circleci.test.ts:181]:
```suggestion
```

**Review by Chriscbr — CHANGES_REQUESTED** [review]

File: src/circleci/circleci.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+  constructor(project: Project, options: CircleCiProps = {}) {
+    super(project);
+    this.options = options;
+    this.orbs = options.orbs ?? {};
+    this.workflows = options.workflows ?? [];
+    this.jobs = options.jobs ?? [];
+    const circleCiEnabled = options.enabled || true;
+    if (circleCiEnabled) {
```
[the remark this answers is not in the document]
Comment (karlderkaefer) [reply src/circleci/circleci.ts]: fixed in b882b4f97a84a417ed1da2a17e52f5dd46432840

**Comment by karlderkaefer** [comment]
@Chriscbr thanks for your review and time, I have addressed all suggestions. I got one question to commit messages, I did not follow the conventional message on every commit, should I rebase and rewrite commit messages, or is that not so important since PRs getting squash merged anyway?

**Pull request merged**